n2v/methods/zmp.py

```diff
@@ -33,174 +35,269 @@ def zmp(self,
             Defined in equation 7 and 8 of reference (1).
+        zmp_functional: str
+            Specifies what functional to use to drive the SCF procedure.
+            Options: {'hartree', 'log', 'exp', 'grad'}
         maxiter: integer, opt
             Maximum number of iterations for SCF calculation
         D_conv: float
             Convergence parameter for density and DIIS error. 
             Default is Psi4's Density convergence parameter: 1e-06
         """
 
-        A = np.array(self.A)
-        S2 = self.S2
-        H = self.T + self.V 
+        #Checks if there is dft grid available. 
+        if hasattr(self.wfn, 'V_potential()') == False:
+            ref = "RV" if self.ref == 1 else "UV"
+            functional = psi4.driver.dft.build_superfunctional("SVWN", restricted=True if self.ref==1 else False)[0]
+            Vpot = psi4.core.VBase.build(self.basis, functional, ref)
+            Vpot.initialize()
+            self.vpot = Vpot
+        else:
+            self.vpot = self.wfn.V_potential()
+
+        # Obtain target density on dft_grid 
+        D0 = self.on_grid_density(grid=None, Da=self.nt[0], Db=self.nt[1],  vpot=self.vpot)
+
+        if zmp_functional.lower() != 'hartree':
+            if self.ref == 1:
+                Da0, Db0 = D0[:,0], D0[:,0]
+            else:
+                Da0, Db0 = D0[:,0], D0[:,1]
+
+            #Calculate kernels
+            if zmp_functional == 'log':
+                Sa0 = np.log(Da0) + 1  
+                Sb0 = np.log(Db0) + 1
+            elif zmp_functional == 'exp':
+                Sa0 = Da0 ** 0.05
+                Sb0 = Db0 ** 0.05
+
+            self.Sa0 = self.dft_grid_to_fock(Sa0, Vpot=self.vpot)
+            self.Sb0 = self.dft_grid_to_fock(Sb0, Vpot=self.vpot)
+
+        vc_global = np.zeros((self.nbf, self.nbf))
+        self.Da = self.nt[0]
+        self.Db = self.nt[1]
 
-        # Trial & Residual Vector Lists
-        state_vectors_a, state_vectors_b = [], []
-        error_vectors_a, error_vectors_b = [], []
+        grid_diff_old = 100.0
 
-        #"Initial Guess for SCF
-        Cocca = self.ct[0]
-        Coccb = self.ct[1]
-        Da = self.nt[0]
-        Db = self.nt[1]
-        D_old = Da.copy()
+#------------->  Iterating over lambdas:
+        for lam_i in lambda_list:
+            self.shift = 0.1 * lam_i
+            Da = self.nt[0]
+            Db = self.nt[1]
 
-        # Obtain slice of target density to test convergence on grid
-        # at the end of scf. 
-        x = np.linspace(-10,10,201)
-        y = np.linspace(-10,10,201)
-        z = np.linspace(-10,10,201)
-        grid = np.concatenate((x[:,None], y[:,None], z[:,None]), axis=1).T
-        density0 = self.on_grid_density(grid, Da=self.nt[0], Db=self.nt[1])
+            Cocca = self.ct[0]
+            Coccb = self.ct[1]
+            D_old = self.nt[0]
+            vc_old = np.zeros_like(Da)
```

Comment:
Probably we don't need to zero vc for each lambda. Instead, vc of old lambda should contribute to vc of current lambda.